### PR TITLE
TST: stats.quantile_test: adjust tests

### PR DESCRIPTION
**Pull request by mdhaber**
Implement some suggestions from scipy/scipy#12680.

### CODE REVIEW

File: scipy/stats/_stats_py.py [hunk 1]
```diff
@@ -9829,15 +9829,15 @@ def quantile_test_iv(x, q, p, alternative):
     if p.ndim != 0 or p >= 1 or p <= 0:
         raise ValueError(message)
 
-    message = "alternative not recognized; \n"
-    "must be 'two-sided', 'less' or 'greater'"
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:9833]: https://github.com/scipy/scipy/pull/12680#discussion_r1230266425

File: scipy/stats/_stats_py.py [hunk 2]
```diff
@@ -9829,15 +9829,15 @@ def quantile_test_iv(x, q, p, alternative):
...
+    alternatives = {'two-sided', 'less', 'greater'}
+    message = f"`alternative` must be one of {alternatives}"
+    if alternative not in alternatives:
         raise ValueError(message)
 
     return x, q, p, alternative
 
 
-def quantile_test(x, q=0, p=0.5, *, alternative='two-sided'):
+def quantile_test(x, *, q=0, p=0.5, alternative='two-sided'):
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:9840]: I'm rethinking which parameters should be keyword only. It is too easy to mix up `p` and `q`, and the defaults are reasonable, so I think this would fit better with where the `scipy.stats` interfaces are headed. Others might even ask for these to be more explicit (`quantile`, `percentile`), but I won't.

File: scipy/stats/tests/test_stats.py [hunk 3]
```diff
@@ -7918,170 +7918,51 @@ def test_alias(self):
...
-    quantile_def = np.random.rand(m)            # quantile definition in [0,1)
-    quantile_val = np.random.rand(m)*100        # quantile value in [0,100)
-    confidence = 0.25*np.random.rand(m)+0.75    # confidence in [0.75,1)
 
     def test_quantile_test_iv(self):
-        
-        x = [1,2,3]
-        p = 0.5
-        q = 0
-        alternative = 'less'
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:7941]: I've removed unnecessary parameters from the calls to emphasize the element being tested.

File: scipy/stats/tests/test_stats.py [hunk 4]
```diff
@@ -7918,170 +7918,51 @@ def test_alias(self):
...
         with pytest.raises(ValueError, match=message):
-            stats.quantile_test(x,q,p,alternative='one-sided')
+            stats.quantile_test(x, alternative='one-sided')
         
         message = "`confidence_level` must be a number between 0 and 1."
         with pytest.raises(ValueError, match=message):
-            stats.quantile_test(x,q,p,alternative=alternative
-                                ).confidence_interval(1)
-
-    def test_confint_naive(self):
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:7968]: It's been a long road, so I would prefer not to review an extra 100 lines. Please let me write a reference implementation, which would be a stronger test because it's from an independent source. Alternatively, I will find a reference implementation.

File: scipy/stats/tests/test_stats.py [hunk 5]
```diff
@@ -7918,170 +7918,51 @@ def test_alias(self):
...
-        assert_equal(public, test_val)
-
-    def test_R_ci_quantile(self):
-        # Using the same test cases as the ci_quantile R function
-        # https://github.com/mayer79/confintr/blob/main/tests/testthat/test-ci_measures_of_location.R#L95
-
+            stats.quantile_test(x).confidence_interval(1)
+
+    # just copying the values you had here, but these should be replaced
+    # with original ones
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:7953]: https://github.com/scipy/scipy/pull/12680#discussion_r1230290452

File: scipy/stats/tests/test_stats.py [hunk 6]
```diff
@@ -7918,170 +7918,51 @@ def test_alias(self):
...
+    @pytest.mark.parametrize('p, alpha, lb, ub',
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:7954]: `parametrize` makes things much more compact.

File: scipy/stats/tests/test_stats.py [hunk 7]
```diff
@@ -7918,170 +7918,51 @@ def test_alias(self):
...
+                             [[0.4, 0.95, 1.349859, 1.648721],
+                              [0.5, 0.9, 1.506818, 1.803988]])
+    def test_R_ci_quantile(self, p, alpha, lb, ub):
+        # Test against R library `confintr` function `ci_quantile`, e.g.
+        # <insert code here>
         x = np.exp(np.arange(0, 1.01, 0.01))
+        res = stats.quantile_test(x, p=p, alternative='two-sided')
+        assert_allclose(res.confidence_interval(alpha), [lb, ub], rtol=1e-5)
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:7962]: https://github.com/scipy/scipy/pull/12680#discussion_r1230298708

File: scipy/stats/tests/test_stats.py [hunk 8]
```diff
@@ -8092,45 +7973,21 @@ def test_pval_ci_match(self):
         # 
         # And the corresponding statement for the alternative='greater' case.
 
-        # This test may fail if we the CI bound falls on a value present 
-        # multiple times in the sample. We make sure this does not happen by 
-        # generating samples via permutations rather than pure random.
-
-        for xi in x:
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:8098]: I misunderstood what was going on:
https://github.com/scipy/scipy/pull/12680#discussion_r1230401636

But I think you'll find the suggested version is a bit simpler. If we want to test for different size samples, we can add, e.g.,  `@pytest.mark.parametrize('size', [23, 85, 129])`.  Where the numbers are arbitrary. I don't think we need 100 of them, though.

File: scipy/stats/tests/test_stats.py [hunk 9]
```diff
@@ -7918,170 +7918,51 @@ def test_alias(self):
...
     including the computation of confidence intervals
     """
-    
-    np.random.seed(0)
-    m = 100                                     # number of tests
-    N = np.random.randint(10, 100, size=m)      # num of samples
-    x = []                                      # sorted samples
-    for i in range(m):
-        num = N[i]
-        x.append(np.sort(np.random.permutation(1000)[:num]))
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:7931]: If you're being careful to ensure that the sample has no tied observations, why not just sample from a continuous distribution? You can still use it to test the edge case that `q` is equal to one of the observations. This would have probability 0 in practice, but you can create a test in which it happens and I think it would test the same thing, effectively. 

(In either case, we also need a test in which observations *are* repeated *and* the quantile is at one of the observations, which is an important edge case.)

**Review by mdhaber — COMMENTED** [review]
Sorry I've been a bit heavy-handed with the simplifications, but this has been going on for a while, so I'd like to get this thing merged as smoothly as possible. Here are some explanations.
